rust/arrow/src/memory.rs

```diff
@@ -135,6 +135,10 @@ const FALLBACK_ALIGNMENT: usize = 1 << 6;
 /// If you use allocation methods shown here you won't have any problems.
 const BYPASS_PTR: NonNull<u8> = unsafe { NonNull::new_unchecked(ALIGNMENT as *mut u8) };
 
+#[cfg(feature = "memory-check")]
+// If this number is not zero after all objects have been `drop`, there is a memory leak
+pub static mut ALLOCATIONS: i32 = 0;
```

Comment:
Just FTR, perhaps you can use a `AtomicUsize` to make this multi-thread-safe?

Reply:
I think so: I was just not very confident with the note

> Note: This type is only available on platforms that support atomic loads and stores of usize.

which I do not know how many / which platforms we would risk with this.

I am also not sure if we should just remove the feature gate and use it. It certainly makes the CI easier, as we do not have to compile it again. Any thoughts @nevi-me @paddyhoran ?

 @pitrou did you ever profiled the C++ code with and without the allocation counter (is it atomic there?)?

Reply:
It's atomic in C++. It's always compiled in, and I don't think we've profiled without it. My general intuition is that an atomic increment will be faster than the allocation itself (an uncontended atomic increment will be a couple nanoseconds at worse?). Also, it sounds unlikely that an application would do many buffer allocations per second.

Reply:
Oh, and by the way, this is a very interested read:
https://travisdowns.github.io/blog/2020/07/06/concurrency-costs.html